scipy/interpolate/_isotonic.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import annotations
+from typing import TYPE_CHECKING
+
+import numpy as np
+
+from scipy.optimize import isotonic_regression
+
+if TYPE_CHECKING:
+    import numpy.typing as npt
+
+
+__all__ = ["IsotonicInterpolator"]
+
+
+# TODO: Should this inherit from _Interpolator1D?
+class IsotonicInterpolator():
+    """Nonparametric isotonic interpolation via PAVA.
+
+    This class provides nonparametric monotonically increasing interpolation
+    via the pool adjacent violators algorithm (PAVA).
+
+    This interpolator first sorts the values of ``y`` according to the order
+    given by ``x``. It then fits the isotonic regression, see
+    ``scipy.optimize.isotonic_regression`` for details. Afterwards it
+    constructs a linear interpolation of the solution of the isotonic
+    regression.
+
+    Parameters
+    ----------
+    x : (N,) array_like or None
+        A 1-D array of real values according to which ``y`` will be sorted.
+    y : (N,) array_like
+        Response variable.
+    weights : (N,) array_like or None
+        Case weights.
+    increasing : bool
+
+    Attributes
+    ----------
+    x_ : array_like
+        Increasing threshold values of ``x`` that define the blocks or pools
+        or knot positions.
+    y_ : array_like
+        Estimated isotonic regression values for all blocks.
+    x_is_numeric : bool
+        True if ``x_`` is numeric.
+
+    Methods
+    -------
+    __call__
+
+    See Also
+    --------
+    scipy.optimize.isotonic_regression : Solves the isotonic regression via PAVA.
+    PchipInterpolator : PCHIP 1-D monotonic cubic interpolator.
+    """
+    def __init__(
+            self,
+            x: npt.ArrayLike,
```

Comment:
Documented as optional but typed as required. Best keep the documentation in line.

Reply:
Does `or None` in the docstring mean optional in scipy?

Reply:
Not even sure what else can it mean.

Reply:
It could mean that `None` is a viable option which can be set. I often find an explicit `"optional"` in scipy docstrings.

Reply:
Yes, `array_like, optional` means it's either an array_like or None. However, the annotation `x: npt.ArrayLike` does not allow None.